server/routes/users.js

```diff
@@ -3,167 +3,473 @@ var express = require('express');
+      }
+    })
+    .catch(err => { return next(err) });
 });
 
 // GET all user emails.
 router.get('/emails', function(req, res, next) {
-  db.query('SELECT email FROM swe_user', [], (err, result) => {
-    if (err) return next(err);
-    res.send(result.rows);
-  });
+  knex('swe_user').select('email')
+    .then(result => {
+      if (result.length) {
+        res.json(result);
+      } else {
+        util.throwError(404, 'No user emails found');
+      }
+    })
+    .catch(err => { return next(err) });
 });
  
 // GET all user university IDs.
 router.get('/ids', function(req, res, next) {
-  db.query('SELECT university_id FROM swe_user;', [], (err, result) => {
-    if (err) return next(err);
-    res.send(result.rows);
-  });
+  knex('swe_user').select('university_id')
+    .then(result => {
+      if (result.length) {
+        res.json(result);
+      } else {
+        util.throwError(404, 'No user university IDs found');
+      }
+    })
+    .catch(err => { return next(err) });
+});
+
+// Add a user
+router.post('/register', function(req, res, next) {
+  const diet_id = req.query.diet_id;
+  const occupation_id = req.query.occupation_id;
+  const position_id = req.query.position_id;
+  const company_id = req.query.company_id;
+  const major_id = req.query.major_id;
+  const rank = req.query.rank;
+
+  knex.transaction(function(trx) { 
+  values =
+  {
+    first_name: req.query.first_name,
+    last_name: req.query.last_name,
+    password: req.query.password,
+    email: req.query.email,
+    phone: req.query.phone,
+    university_id: req.query.university_id,
+    is_admin: req.query.is_admin,
+  }
+  return trx
+  .insert(values, 'id')
+  .into('swe_user')
+  .transacting(trx)
+  .then(function(ids){
+
+    diet_values = {
+      user_id: ids[0],
+      diet_id: diet_id
+    }
+    occupation_values = {
+      user_id: ids[0],
+      occupation_id: occupation_id
+    }
+    position_values = {
+      user_id: ids[0],
+      position_id: position_id
+    }
+    major_values = {
+      user_id:  ids[0],
+      major_id: major_id,
+    } 
+    company_rank_values = {
+      user_id:  ids[0],
+      company_id: company_id,
+      rank: rank
+    }
+
+    inserts = []
+
+    if(diet_id){
+      inserts.push({
+        values: diet_values,
+        table_name: 'user_diet'
+      });
+    }
+    if(occupation_id){
+      inserts.push({
+        values: occupation_values,
+        table_name: 'user_occupation'
+      });
+    }
+    if(position_id){
+      inserts.push({
+        values: position_values,
+        table_name: 'user_position'
+      });
+    }
+    if(major_id){
+      inserts.push({
+        values: major_values,
+        table_name: 'user_major'
+      });
+    }
+    if(company_id && rank){
+      inserts.push({
+        values: company_rank_values,
+        table_name: 'user_company_rank'
+      });
+    }
+    console.log(ids[0]); 
+    return Promise.map(inserts, function(insert_obj){
```

Comment:
This seems like an overly complicated way to form this query? Let's try simplifying this so it's more readable. I like the concept, but this adds like 5-10 lines per insert rather than 1-3 lines? maybe?

Reply:
fixed, changed to async/await